storage/src/adb/store/mod.rs

```diff
@@ -176,20 +171,34 @@ pub trait Db<K: Array, V: Codec> {
         }
     }
 
-    /// Creates a new key-value pair in the db. The operation is reflected in the snapshot, but will
-    /// be subject to rollback until the next successful `commit`. Returns true if the key was
-    /// created, false if it already existed.
-    fn create(&mut self, key: K, value: V) -> impl Future<Output = Result<bool, Error>>;
-
-    /// Delete `key` and its value from the db. Deleting a key that already has no value is a no-op.
-    /// The operation is reflected in the snapshot, but will be subject to rollback until the next
-    /// successful `commit`. Returns true if the key was deleted, false if it was already inactive.
+    /// Delete `key` and its value from the store. Deleting a key that already has no value is a
+    /// no-op. Returns true if the key was deleted, false if it was already inactive.
     fn delete(&mut self, key: K) -> impl Future<Output = Result<bool, Error>>;
+}
+
+/// A prunable log of operations backing the key-value store.
+pub trait Log {
+    /// Prune operations prior to `prune_loc`.
+    fn prune(&mut self, prune_loc: Location) -> impl Future<Output = Result<(), Error>>;
+
+    /// The number of operations that have been applied to this store, including those that have been
+    /// pruned and those that are not yet committed.
+    fn op_count(&self) -> Location;
+
+    /// Return the inactivity floor location. This is the location before which all operations are
+    /// known to be inactive. Operations before this point can be safely pruned.
+    fn inactivity_floor_loc(&self) -> Location;
+}
+
+/// A key-value store backed by a log that can be committed and persisted.
+pub trait PersistedKeyValueStore<K: Array, V: Codec>: KeyValueStore<K, V> {
```

Comment:
thought I left this suggestion before but I don't see it, so apologies if this is duplicate:

Should this be Persistable instead of Persisted?  (It's not persisted if it's dirty...)